GroundedFunctionNetwork creation and analysis

pauldhein opened the pull request:
## Changes
- Implemented a `GroundedFunctionNetwork` class that can be constructed from Fortran and serves as a real exectuable representation of a `GrFN`.
- Implemented a `run()` method that allows the a `GrFN` to be run with either single or batched input examples
- Implemented a `to_ProgramAnalysisGraph` method that allows us to view the variable connections in the GrFN
- Implemented visualization methods that allow us to view a full GrFN
- Changed the implementation of Sobol analysis to work with GrFN as a single function
- Implemented FAST sensitivity analysis for quickly getting `S1` and `ST` indices
- Removed large amounts of file and folder bloat from `delphi/GrFN` and `delphi/analysis`

## Requirements before merge
1. Complete testing with `crop_yield.f` (Requires #211 )
2. Complete visualization testing with a module that contains sub-modules (Requires `1`)

## Additions coming soon
- Surface plot function from GrFN evaluation
- Testing ForwardInfluenceBlanket generation from GrFN (Requires translated `PETASCE`)

adarshp commented on delphi/GrFN/networks.py:
```diff
@@ -0,0 +1,383 @@
+
+class GroundedFunctionNetwork(nx.DiGraph):
+    """
+    Stub for the Grounded Function Network class
```
No longer a stub :)

adarshp commented on delphi/GrFN/networks.py:
```diff
@@ -0,0 +1,383 @@
+    def traverse_nodes(self, node_set, depth=0):
+        """BFS traversal of nodes that returns name traversal as large string."""
+        tab = "  "
+        result = ""
```
Construct a list of strings and then concatenate the list, instead of doing `+=` (see https://docs.python-guide.org/writing/structure/#mutable-and-immutable-types).

adarshp commented on delphi/GrFN/networks.py:
```diff
@@ -0,0 +1,383 @@
+    @classmethod
+    def from_dict(cls, data: Dict, lambdas):
+        """
+        Builds a GrFN object from a set of extracted function data objects and
```
Can you change the docstring style to the Google style (that's what is being used in the rest of the repo)? Here's an example from https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html#example-google

```python
def function_with_pep484_type_annotations(param1: int, param2: str) -> bool:
    """Example function with PEP 484 type annotations.

    Args:
        param1: The first parameter.
        param2: The second parameter.

    Returns:
        The return value. True for success, False otherwise.

    """
```

pauldhein commented:
## Update

#### Fixes
- Addressed most of Adarsh's concerns. @adarshp please re-review the code
- Fixed issues with running GrFN with generalized matrix inputs
- Now hav a test with a modified `crop_yield` lambdas + JSON

#### Additions
- Now have `RBD-FAST` sensitivity analysis
- Now computing function heap for GrFN at creation time instead of runtime
- Added max S2 sensitivity surface computation
- Added ability to construct a runnable `ForwardInfluenceBlanket` from a `GroundedFunctionNetwork`

#### Todo's
- Test the running of `ForwardInfluenceBlanket`'s on `PETPT` and `PETASCE`
- Test `ForwardInfluenceBlanket` visualization
- Test sensitivity surface visualization
- Ensure that GrFNs work properly with the generated `crop_yield`
- Integrate subgraph visualization into GrFN class for webapp

adarshp commented on tests/test_analysis.py:
```diff
@@ -1,100 +1,167 @@
+import json
+import sys
+
+# import delphi.analysis.sensitivity.variance_methods as methods
```
Ok for now if you're still testing things out, but we should remove old commented-out code before merging into master.

adarshp commented on delphi/GrFN/utils.py:
```diff
@@ -0,0 +1,48 @@
+    return timed
+
+
+class NodeType(Enum):
```
I assume this Enum is still on its way out?

adarshp commented on delphi/GrFN/networks.py:
```diff
@@ -0,0 +1,640 @@
+
+class ForwardInfluenceBlanket(nx.DiGraph):
+    """
+    This class takes a network and a list of a shared nodes between the input network and a secondary network. From this list a shared nodes and blanket network is created including all of the nodes between any input/output pair in the shared nodes, as well as all nodes required to blanket the network for forward influence. This class itself becomes the blanket and inherits from the NetworkX DiGr ...
```
Line wrapping
